File: Leetcode-Python/FlattenBinaryTreeToLinkedList.py
"""
Given a binary tree, flatten it to a linked list in-place.

For example,
Given

         1
        / \
       2   5
      / \   \
     3   4   6

The flattened tree should look like:
   1
    \
     2
      \
       3
        \
         4
          \
           5
            \
             6

"""
from TreeNode import TreeNode


class Solution:
    def __init__(self):
        self.__pre = None

    # @param {TreeNode} root
    # @return {void} Do not return anything, modify root in-place instead.
    def flatten(self, root):
        if not root:
            return
        temp = root.right
        if self.__pre:
            self.__pre.left, self.__pre.right = None, root
        self.__pre = root
        self.flatten(root.left)
        self.flatten(temp)

if __name__ == '__main__':
    n1 = TreeNode(1)
    n2 = TreeNode(2)
    n3 = TreeNode(3)
    n4 = TreeNode(4)
    n5 = TreeNode(5)
    n6 = TreeNode(6)

    n1.left = n2
    # n1.right = n5
    n2.left = n3
    # n2.right = n4
    # n5.right = n6

    head = Solution().flatten(n1)

File: Leetcode-Java/FlattenBinaryTreeToLinkedList.java
/**
 * Created by yhf on 15/5/8.
 */

/*
Given a binary tree, flatten it to a linked list in-place.

For example,
Given
         1
        / \
       2   5
      / \   \
     3   4   6

The flattened tree should look like:

   1
    \
     2
      \
       3
        \
         4
          \
           5
            \
             6
*/

/*
* 1. 先把本次访问right的存下来
* 2. lastVisit.left设为null，lastVisit.right设为root
* 3. 递归，先序遍历
* */

public class FlattenBinaryTreeToLinkedList {
    private static TreeNode lastVisit = null;

    public static void flatten(TreeNode root) {
        if (root == null) {
            return;
        }
        TreeNode saveRight = root.right;
        if (lastVisit != null) {
            lastVisit.left = null;
            lastVisit.right = root;
        }
        lastVisit = root;
        flatten(root.left);
        flatten(saveRight);
    }

    public static void main(String[] args) {
        TreeNode n1 = new TreeNode(1);
        TreeNode n2 = new TreeNode(2);
        TreeNode n3 = new TreeNode(3);
        TreeNode n4 = new TreeNode(4);
        TreeNode n5 = new TreeNode(5);
        TreeNode n6 = new TreeNode(6);
        n1.left = n2;
        n2.left = n3;
        n2.right = n4;
        n1.right = n5;
        n5.right = n6;
        flatten(n1);
    }
}
